### PR TITLE
Cache provider

### PR DESCRIPTION
**Pull request by legionus**
New cache provider should join our cache and `docker/distribution` cache in one. The expiration time  now affects it. Also we can prune it.

### CODE REVIEW

File: pkg/dockerregistry/server/app.go [hunk 1]
```diff
@@ -74,6 +75,12 @@ func (app *App) BlobStatter() distribution.BlobStatter {
 	}
 }
 
+func (app *App) CacheProvider(ctx context.Context, options map[string]interface{}) (registrycache.BlobDescriptorCacheProvider, error) {
+	return &cache.CacheProvider{
```
Comment (dmage) [inline pkg/dockerregistry/server/app.go]: Stutter. s/CacheProvider/Provider/

File: pkg/dockerregistry/server/supermiddleware/app_test.go [hunk 2]
```diff
@@ -108,6 +109,12 @@ func (app *testApp) Repository(ctx context.Context, repo distribution.Repository
 	return repo, bdsf, nil
 }
 
+func (app *testApp) CacheProvider(ctx context.Context, options map[string]interface{}) (cache.BlobDescriptorCacheProvider, error) {
+	return &testCacheProvider{
```
Comment (dmage) [inline pkg/dockerregistry/server/supermiddleware/app_test.go]: What are you testing with it?

File: pkg/dockerregistry/server/app.go (hunk 1 above)
Comment (dmage) [reply pkg/dockerregistry/server/app.go]: Oops, it's upstream naming :(

**Comment by dmage** [comment]
@legionus please update the commit for the vendor directory (s/<carry>/2580/).

**Comment by legionus** [comment]
@dmage Done

**Pull request merged**